mds/src/util.rs

```diff
@@ -0,0 +1,186 @@
+    const MAT_1: [u64; N] = rotate_right(MATRIX_CIRC_MDS_12_SML, 1);
+    output[1] = F::z_linear_combination_sml(MAT_1, &input);
+    const MAT_2: [u64; N] = rotate_right(MATRIX_CIRC_MDS_12_SML, 2);
+    output[2] = F::z_linear_combination_sml(MAT_2, &input);
+    const MAT_3: [u64; N] = rotate_right(MATRIX_CIRC_MDS_12_SML, 3);
+    output[3] = F::z_linear_combination_sml(MAT_3, &input);
+    const MAT_4: [u64; N] = rotate_right(MATRIX_CIRC_MDS_12_SML, 4);
+    output[4] = F::z_linear_combination_sml(MAT_4, &input);
+    const MAT_5: [u64; N] = rotate_right(MATRIX_CIRC_MDS_12_SML, 5);
+    output[5] = F::z_linear_combination_sml(MAT_5, &input);
+    const MAT_6: [u64; N] = rotate_right(MATRIX_CIRC_MDS_12_SML, 6);
+    output[6] = F::z_linear_combination_sml(MAT_6, &input);
+    const MAT_7: [u64; N] = rotate_right(MATRIX_CIRC_MDS_12_SML, 7);
+    output[7] = F::z_linear_combination_sml(MAT_7, &input);
+    const MAT_8: [u64; N] = rotate_right(MATRIX_CIRC_MDS_12_SML, 8);
+    output[8] = F::z_linear_combination_sml(MAT_8, &input);
+    const MAT_9: [u64; N] = rotate_right(MATRIX_CIRC_MDS_12_SML, 9);
+    output[9] = F::z_linear_combination_sml(MAT_9, &input);
+    const MAT_10: [u64; N] = rotate_right(MATRIX_CIRC_MDS_12_SML, 10);
+    output[10] = F::z_linear_combination_sml(MAT_10, &input);
+    const MAT_11: [u64; N] = rotate_right(MATRIX_CIRC_MDS_12_SML, 11);
+    output[11] = F::z_linear_combination_sml(MAT_11, &input);
+
+    output
+}
+
+/// Given the first row of a circulant matrix, return the first column
+/// of that circulant matrix. For example, v = [0, 1, 2, 3, 4, 5],
+/// then output = [0, 5, 4, 3, 2, 1], i.e. the first element is the
+/// same and the other elements are reversed.
+///
+/// This is useful to prepare a circulant matrix for input to an FFT
+/// algorithm, which expects the first column of the matrix rather
+/// than the first row (as we normally store them).
+///
+/// NB: The algorithm is inefficient but simple enough that this
+/// function can be declared `const`, and that is the intended context
+/// for use.
+pub(crate) const fn first_row_to_first_col<const N: usize>(v: &[u64; N]) -> [u64; N] {
+    let mut output = [0u64; N];
+    output[0] = v[0];
+    let mut i = 1;
+    loop {
+        if i >= N {
+            break;
+        }
+        output[i] = v[N - i];
+        i += 1;
+    }
+    output
+}
+
+/// Use the convolution theorem to calculate the product of the given
+/// circulant matrix and the given vector. The circulant matrix must
+/// be specified by its first *column*, not its first row. If you have
+/// the row as an array, you can obtain the column with `first_row_to_first_column()`.
+pub(crate) fn apply_circulant_fft<F: TwoAdicField, const N: usize, FFT: TwoAdicSubgroupDft<F>>(
+    fft: FFT,
+    column: [u64; N],
+    input: &[F; N],
+) -> [F; N] {
+    let column = column.map(F::from_canonical_u64).to_vec();
+    let matrix = fft.dft_batch(RowMajorMatrix::new_col(column));
```

Comment:
Hopefully this will normally happen at compile time? Maybe just add `#[inline]` to encourage it?